src/runtime/stream.rs

```diff
@@ -0,0 +1,293 @@
+use std::{
+    net::{SocketAddr, ToSocketAddrs},
+    ops::DerefMut,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use crate::{
+    cmap::conn::StreamOptions,
+    error::{ErrorKind, Result},
+    options::StreamAddress,
+};
+
+const DEFAULT_CONNECT_TIMEOUT: Duration = Duration::from_secs(10);
+
+/// A runtime-agnostic async stream possibly using TLS.
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+pub(crate) enum AsyncStream {
+    /// A basic TCP connection to the server.
+    Tcp(AsyncStreamInner),
+
+    /// A TLS connection over TCP.
+    Tls(tokio_rustls::client::TlsStream<AsyncStreamInner>),
+}
+
+/// A runtime-agnostic async stream.
+#[derive(Debug)]
+pub(crate) enum AsyncStreamInner {
+    /// Wrapper around `tokio::net:TcpStream`.
+    #[cfg(feature = "tokio-runtime")]
+    Tokio(tokio::net::TcpStream),
+
+    /// Wrapper around `async_std::net::TcpStream`.
+    #[cfg(feature = "async-std-runtime")]
+    AsyncStd(async_std::net::TcpStream),
+}
+
+#[async_trait::async_trait]
+trait AsyncConnect
```

Comment:
This trait is needed in order to avoid duplicating the code that implements fallback for the various addresses returned for a domain.